File: 04.median-of-two-sorted-arrays/Median.py


def find_median(arr : list[int]) -> float:
    k = len(arr) // 2
    if len(arr) % 2 == 1:
        return arr[k]
    return (arr[k-1] + arr[k]) / 2


def merge(one: list[int], two: list[int]) -> list[int]:
    i, j = 0, 0
    whole = []
    for _ in range(len(one) + len(two)):
        if j == len(two) or (i < len(one) and j < len(two) and one[i] <= two[j]):
            whole.append(one[i])
            i += 1
            continue
        if i == len(one) or (i < len(one) and j < len(two) and one[i] >= two[j]):
            whole.append(two[j])
            j += 1
    return whole

def calc_median(one:list[int], two:list[int]) -> float:
    return find_median(merge(one, two))

if __name__ == '__main__':
    print(calc_median([], [7, 8, 9, 10]))

File: 04.median-of-two-sorted-arrays/Median.go
package main

import "fmt"

func calcMedian(one, two []int) float32 {
	whole := merge(one, two)
	return findMedian(whole)
}

func merge(one, two []int) []int {
	whole := make([]int, len(one)+len(two))
	i, j := 0, 0

	for index := 0; index < (len(one) + len(two)); index++ {
		if j == len(two) || (i < len(one) && j < len(two) && one[i] < two[j]) {
			whole[index] = one[i]
			i++
			continue
		}

		if i == len(one) || (i < len(one) && j < len(two) && one[i] >= two[j]) {
			whole[index] = two[j]
			j++
		}
	}

	return whole
}

func findMedian(arr []int) float32 {
	k := len(arr) / 2
	if len(arr)%2 == 1 {
		return float32(arr[k])
	} else {
		return float32(arr[k]+arr[k-1]) / 2.0
	}
}

func solveProblem(one, two []int) {
	whole := merge(one, two)
	median := calcMedian(one, two)
	fmt.Printf("array one: %v\n", one)
	fmt.Printf("array two: %v\n", two)
	fmt.Printf("array whole: %v\n", whole)
	fmt.Printf("median: %f\n", median)

}

func main() {
	solveProblem([]int{2, 4}, []int{7, 8, 9})

	solveProblem([]int{2, 4, 8}, []int{7, 8, 9})
}
